File: CIA - 1/12_best_first_search.py
import networkx as nx
import matplotlib.pyplot as plt
import heapq

def best_first_search(graph, start, goal, heuristic):
    frontier = [(heuristic[start], start, [start])]
    visited = set()

    while frontier:
        _, node, path = heapq.heappop(frontier)

        if node == goal:
            return path

        visited.add(node)

        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                heapq.heappush(frontier, (heuristic[neighbor], neighbor, path + [neighbor]))

    print("No path found.")
    return None

# Input graph details and heuristic values
num_vertices = int(input("Enter the number of vertices: "))
edges = []
for _ in range(int(input("Enter the number of edges: "))):
    edge = input("Enter edge (format: source target weight): ").split()
    edges.append((edge[0], edge[1], int(edge[2])))

# Heuristic values for each node
heuristic = {}
for _ in range(num_vertices):
    vertex, value = input("Enter vertex and heuristic value (format: vertex value): ").split()
    heuristic[vertex] = int(value)

start_vertex = input("Enter the start vertex: ")
goal_vertex = input("Enter the goal vertex: ")

# Create a weighted graph using networkx
G = nx.Graph()
for edge in edges:
    G.add_edge(edge[0], edge[1], weight=edge[2])

# Perform Best First Search
path = best_first_search(G, start_vertex, goal_vertex, heuristic)

# Draw the initial graph
pos = nx.spring_layout(G)
plt.figure(figsize=(10, 7))
nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
edge_labels = nx.get_edge_attributes(G, 'weight')
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels)
plt.title("Initial Graph")
plt.show()

# If a path is found, display the path
if path:
    path_edges = list(zip(path, path[1:]))
    plt.figure(figsize=(10, 7))
    nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
    nx.draw_networkx_edges(G, pos, edgelist=path_edges, edge_color='r', width=2)
    nx.draw_networkx_nodes(G, pos, nodelist=path, node_color='lightgreen')
    plt.title(f"Best First Search Path: {path}")
    plt.show()
else:
    print("No path found.")

File: CIA - 1/12_best_first_search.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

#define MAX 100

int adjMatrix[MAX][MAX];
int visited[MAX];
int numVertices;
int heuristic[MAX]; // Heuristic values for each vertex

// Function to get the minimum heuristic node that has not been visited
int getMinHeuristicNode() {
    int min = INT_MAX;
    int minIndex = -1;

    for (int i = 0; i < numVertices; i++) {
        if (!visited[i] && heuristic[i] < min) {
            min = heuristic[i];
            minIndex = i;
        }
    }

    return minIndex;
}

// Best-First Search algorithm
void bestFirstSearch(int start, int goal) {
    int currentNode = start;
    printf("Path: %c ", currentNode + 'A');
    visited[currentNode] = 1;

    while (currentNode != goal) {
        int nextNode = getMinHeuristicNode();

        if (nextNode == -1) {
            printf("\nNo path found to goal!\n");
            return;
        }

        printf("-> %c ", nextNode + 'A');
        visited[nextNode] = 1;

        currentNode = nextNode;
    }

    printf("\nGoal reached: %c\n", goal + 'A');
}

int getVertexIndex(char vertices[], char vertex) {
    for (int i = 0; i < numVertices; i++) {
        if (vertices[i] == vertex) {
            return i;
        }
    }
    return -1;
}

int main() {
    int numEdges;
    char vertices[MAX];

    printf("Enter the number of vertices: ");
    scanf("%d", &numVertices);

    printf("Enter the names of the vertices:\n");
    for (int i = 0; i < numVertices; i++) {
        scanf(" %c", &vertices[i]);
    }

    printf("Enter the number of edges: ");
    scanf("%d", &numEdges);

    // Initialize adjacency matrix and visited array
    for (int i = 0; i < numVertices; i++) {
        visited[i] = 0;
        for (int j = 0; j < numVertices; j++) {
            adjMatrix[i][j] = 0;
        }
    }

    printf("Enter the edges (source destination weight):\n");
    for (int i = 0; i < numEdges; i++) {
        char s, d;
        int weight;
        scanf(" %c %c %d", &s, &d, &weight);
        int source = getVertexIndex(vertices, s);
        int destination = getVertexIndex(vertices, d);

        if (source == -1 || destination == -1) {
            printf("Invalid vertex name\n");
            return -1;
        }

        adjMatrix[source][destination] = weight;
        adjMatrix[destination][source] = weight;  // Assuming undirected graph
    }

    // Input heuristics
    printf("Enter heuristic values for each vertex:\n");
    for (int i = 0; i < numVertices; i++) {
        printf("Heuristic for %c: ", vertices[i]);
        scanf("%d", &heuristic[i]);
    }

    char startVertex, goalVertex;
    printf("Enter the start vertex: ");
    scanf(" %c", &startVertex);

    printf("Enter the goal vertex: ");
    scanf(" %c", &goalVertex);

    int start = getVertexIndex(vertices, startVertex);
    int goal = getVertexIndex(vertices, goalVertex);

    if (start == -1 || goal == -1) {
        printf("Invalid start or goal vertex\n");
        return -1;
    }

    // Run Best-First Search
    bestFirstSearch(start, goal);

    return 0;
}
